include/boost/geometry/algorithms/detail/is_valid/pointlike.hpp

```diff
@@ -42,14 +45,26 @@ struct is_valid<Point, point_tag>
 // (have identical coordinate values in X and Y)
 //
 // Reference: OGC 06-103r4 (6.1.5)
-template <typename MultiPoint>
-struct is_valid<MultiPoint, multi_point_tag>
+template <typename MultiPoint, bool AllowEmptyMultiGeometries>
+struct is_valid<MultiPoint, multi_point_tag, AllowEmptyMultiGeometries>
 {
-    static inline bool apply(MultiPoint const& /*multipoint*/)
+    template <typename VisitPolicy>
+    static inline bool apply(MultiPoint const& multipoint,
+                             VisitPolicy& visitor)
     {
-        // we allow empty multi-geometries, so an empty multipoint
-        // is considered valid
-        return true;
+        if (AllowEmptyMultiGeometries || boost::size(multipoint) > 0)
+        {
+            // we allow empty multi-geometries, so an empty multipoint
+            // is considered valid
+            visitor.template apply<no_failure>();
+            return true;
+        }
+        else
+        {
+            // we do not allow an empty multipoint
+            visitor.template apply<failure_few_points>();
+            return false;
```

Comment:
What I can do it is to have the visitor's apply method return the boolean. If that is what you mean it is easy to do. I already considered doing so, but I decided to avoid it: the drawback that I see is that it "complicates" the interface of the visitor. I would prefer the visitor to just take info from the algorithm passively, rather than actively participate in computing the result. This way we can avoid potential "malicious" policies that disregard the failure value and return whatever....

What I think (I have to double-check this, since I think I already tried it) I cannot do is to eliminate the `if (...)` statement. When I call the visitor I need the exact precise failure type at compile time.